### PR TITLE
sync 2.2  Fed Stats changes back to Dev branch [ skip ci]

### PR DESCRIPTION
**Pull request by chesterxgchen**

### CODE REVIEW

File: tests/unit_test/app_common/workflow/statistics_controller_test.py [hunk 1]
```diff
@@ -81,4 +102,15 @@ def test_prepare_input(self):
 
         xs = self.stats_controller._prepare_inputs(SC.STATS_2nd_METRICS, None)
         assert xs[SC.METRIC_TASK_KEY] == SC.STATS_2nd_METRICS
-        assert xs[SC.STATS_TARGET_METRICS].sort() == SC.ordered_metrics[SC.STATS_2nd_METRICS].sort()
+        assert fobs.loads(xs[SC.STATS_TARGET_METRICS]).sort() == SC.ordered_metrics[SC.STATS_2nd_METRICS].sort()
```
Comment (nvidianz) [inline tests/unit_test/app_common/workflow/statistics_controller_test.py:105]: list.sort() returns None so this comparison may not work.

File: tests/unit_test/app_common/workflow/statistics_controller_test.py [hunk 2]
```diff
@@ -29,7 +32,7 @@ def setup_class(cls):
             "stddev": {},
             "histogram": {"*": {"bins": 10}, "Age": {"bins": 5, "range": [0, 120]}},
         }
-        cls.stats_controller = MockGlobalStatistics(metric_configs=metric_configs, writer_id="")
+        cls.stats_controller = MockStatisticsController(metric_configs=metric_configs, writer_id="")
 
     @classmethod
     def teardown_class(cls):
```
Comment (YuanTingHsieh) [inline tests/unit_test/app_common/workflow/statistics_controller_test.py:38]: maybe we should remove unused methods.

File: nvflare/app_common/workflows/statistics_controller.py [hunk 3]
```diff
@@ -226,25 +289,39 @@ def _combine_all_metrics(self):
...
                         )
 
         return result
 
     def _get_target_metrics(self, ordered_metrics) -> List[MetricConfig]:
+        return StatisticsController._get_target_metrics2(self.metric_configs, ordered_metrics)
+
+    @staticmethod
+    def _get_target_metrics2(metric_configs: dict, ordered_metrics: list) -> List[MetricConfig]:
         # make sure the execution order of the metrics calculation
```
Comment (YuanTingHsieh) [inline nvflare/app_common/workflows/statistics_controller.py:321]: why are we having the `_get_target_metrics2` method?

File: nvflare/app_common/workflows/statistics_controller.py [hunk 4]
```diff
@@ -273,17 +350,70 @@ def _prepare_inputs(self, metric_task: str, fl_ctx: FLContext) -> Shareable:
...
         return inputs
 
+    def _wait_for_all_results(
+        self,
+        result_wait_timeout: float,
+        requested_client_size: int,
+        client_metrics: dict,
+        sleep_time: float = 1,
+        abort_signal=None,
+    ) -> bool:
```
Comment (YuanTingHsieh) [inline nvflare/app_common/workflows/statistics_controller.py:377]: suggest make this method either static method or function outside of class.

Can just pass in the self.logger

File: tests/unit_test/app_common/workflow/mock_statistics_controller.py [hunk 5]
```diff
@@ -21,9 +22,9 @@
 from nvflare.app_common.workflows.statistics_controller import StatisticsController
 
 
-class MockGlobalStatistics(StatisticsController):
+class MockStatisticsController(StatisticsController):
```
Comment (YuanTingHsieh) [inline tests/unit_test/app_common/workflow/mock_statistics_controller.py:25]: I suggest we remove this class if we are mocking every method as pass

File: tests/unit_test/app_common/workflow/statistics_controller_test.py [hunk 6]
```diff
@@ -65,10 +68,28 @@ def test_target_metrics(self):
...
+        client_metrics = {
+            "count": {"site-1": {}},
+            "mean": {"site-2": {}},
+            "sum": {"site-3": {}},
+            "stddev": {"site-4": {}},
+        }
+        import time
+
+        t0 = time.time()
+        self.stats_controller._wait_for_all_results(0.5, 3, client_metrics, 0.1)
```
Comment (YuanTingHsieh) [inline tests/unit_test/app_common/workflow/statistics_controller_test.py:83]: if this method is static method or function, we don't need a class to test it.

File: nvflare/app_common/workflows/statistics_controller.py (hunk 3 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/statistics_controller.py:321]: if they are the same we could just have 


```
@staticmethod
def _get_target_metrics(metric_configs: dict, ordered_metrics: list) -> List[MetricConfig]:

```

File: tests/unit_test/app_common/workflow/statistics_controller_test.py (hunk 6 above)
Comment (YuanTingHsieh) [reply tests/unit_test/app_common/workflow/statistics_controller_test.py:83]: but we could do that in follow up PRs.

File: nvflare/app_common/workflows/statistics_controller.py (hunk 4 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/statistics_controller.py:377]: but we could do that in follow up PRs.

**Review by nvidianz — APPROVED** [review]

**Pull request merged**